### PR TITLE
Refine DropoutOp

### PR DESCRIPTION
**Pull request by sneaxiy**

### CODE REVIEW

File: paddle/fluid/operators/dropout_op.cu [hunk 1]
```diff
@@ -23,35 +23,30 @@ limitations under the License. */
...
 namespace operators {
 
-template <typename T>
-__global__ void RandomGenerator(const size_t n, const int seed,
-                                const float dropout_prob, const T* src,
-                                T* mask_data, T* dst) {
-  thrust::minstd_rand rng;
-  rng.seed(seed);
-  thrust::uniform_real_distribution<float> dist(0, 1);
+inline __host__ __device__ size_t HashCombine(size_t seed, size_t idx) {
```
Comment (reyoung) [inline paddle/fluid/operators/dropout_op.cu]: `inline static __device__`

**Review by reyoung — DISMISSED** [review]
Basic LGTM, however, does the transformer or resnet produce the similar loss?

File: paddle/fluid/operators/dropout_op.cu [hunk 2]
```diff
@@ -23,82 +23,52 @@ limitations under the License. */
...
+                                const uint32_t int_dropout_prob, const T *src,
+                                MaskType *mask_data, T *dst) {
+  size_t idx = blockDim.x * blockIdx.x + threadIdx.x;
   thrust::minstd_rand rng;
-  rng.seed(seed);
-  thrust::uniform_real_distribution<float> dist(0, 1);
-
-  int idx = blockDim.x * blockIdx.x + threadIdx.x;
-  int step_size = 0;
+  rng.seed(HashCombine(seed, idx));
```
Comment (sneaxiy) [inline paddle/fluid/operators/dropout_op.cu]: Maybe we should find a better way to combine `seed` and `idx` to guarantee the independence of random numbers between different CUDA threads.

File: paddle/fluid/operators/dropout_op.h [hunk 3]
```diff
@@ -38,39 +43,61 @@ class CPUDropoutKernel : public framework::OpKernel<T> {
...
+      size_t size = x->numel();
+      DropoutFunctor functor;
+      functor(context.template device_context<DeviceContext>(), x_data, y_data,
+              mask_data, size, dropout_prob, seed);
+
+      if (fix_seed) {
+        seed = HashCombine(
+            seed, static_cast<uint32_t>(static_cast<uint32_t>(-1) *
+                                        static_cast<double>(dropout_prob)));
+        seed = HashCombine(seed, size);
```
Comment (chengduoZH) [inline paddle/fluid/operators/dropout_op.h:80]: Why do you call `HashCombine ` twice here?

File: paddle/fluid/operators/dropout_op.cc [hunk 4]
```diff
@@ -72,27 +106,48 @@ are set equal to their corresponding inputs.
...
+  }
+
+  framework::OpKernelType GetKernelTypeForVar(
+      const std::string& var_name, const framework::Tensor& tensor,  // NOLINT
+      const framework::OpKernelType& expected_kernel_type) const override {
+    // Prevent data transform of Input(SeedIn) from CPUPlace to CUDAPlace
+    return (var_name == "SeedIn" ||
+            var_name == framework::GradVarName("SeedIn") ||
+            var_name == "SeedOut" ||
+            var_name == framework::GradVarName("SeedOut"))
```
Comment (chengduoZH) [inline paddle/fluid/operators/dropout_op.cc]: Why do "SeedIn" and "SeedOut" have gradient?

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -947,22 +948,48 @@ def dropout(x, dropout_prob, is_test=False, seed=None, name=None):
...
+            stop_gradient=True,
+            initializer=Constant(
+                value=seed, force_cpu=True))
+
+        seed_var = default_main_program().current_block().create_var(
+            name=seed_name,
+            shape=[1],
+            dtype=seed_dtype,
+            persistable=True,
+            stop_gradient=True)
```
Comment (chengduoZH) [inline python/paddle/fluid/layers/nn.py]: You can add `stop_gradient=True` for `SeedIn` here.

File: paddle/fluid/operators/dropout_op.h (hunk 3 above)
Comment (sneaxiy) [reply paddle/fluid/operators/dropout_op.h:80]: It has no sense. Just to make seed more random and depend on both `dropout_prob` and `size`.

File: paddle/fluid/operators/dropout_op.cc (hunk 4 above)
Comment (sneaxiy) [reply paddle/fluid/operators/dropout_op.cc]: You are right. They should not have gradient.

File: paddle/fluid/operators/dropout_op.cu [hunk 6]
```diff
@@ -13,92 +13,62 @@ See the License for the specific language governing permissions and
...
+                                MaskType *mask_data, T *dst) {
+  constexpr uint32_t kUInt32Max = static_cast<uint32_t>(-1);
+  size_t idx = blockDim.x * blockIdx.x + threadIdx.x;
 
-  int idx = blockDim.x * blockIdx.x + threadIdx.x;
-  int step_size = 0;
+  thrust::minstd_rand rng;
+  thrust::uniform_int_distribution<uint32_t> dist(0, kUInt32Max - 1);
+  rng.seed(HashCombine(seed, idx));
+  rng.seed(dist(rng));
```
Comment (sneaxiy) [inline paddle/fluid/operators/dropout_op.cu:33]: @reyoung @chengduoZH @JiayiFeng `rng.seed(dist(rng))` is weird here. If I do not add `rng.seed(dist(rng))`, `nan` may occur in some test cases even though the actual `dropout_element_num/total_element_num` is still close to `dropout_prob`. The reason may be that there are lots of continuous zeros in `Mask`. Therefore, maybe it is not a best way to produce random seed by `HashCombine`. See #12639 as an alternative.

**Pull request closed without merging**